### PR TITLE
Implement null_safety rule

### PR DESCRIPTION
**Pull request by asashour**
Implements the first requirement of #2702

I am not sure how best to process `pubspec.yaml`

Should the name be `sound_null_safety`, `unsound_null_safety`, ...?

### CODE REVIEW

File: lib/src/rules/null_safety.dart
Comment (bwilkerson) [inline lib/src/rules/null_safety.dart:34]: Although it isn't strictly a requirement, consider naming this with a verb phrase, something like `enable_null_safety`. We generally try to name lints so that they describe either what _should_ be done or what _shouldn't_ be done.

File: lib/src/rules/null_safety.dart
Comment (bwilkerson) [inline lib/src/rules/null_safety.dart:58]: This regex doesn't work with all of the possibilities. Users are allowed to have whitespace almost anywhere in the line, and we probably need to catch comments that use a major version of `1` (though that might be disallowed in general; I cant remember). We should probably share the regex used by `LanguageVersionOverrideVerifier` (in the analyzer package). Unfortunately that class isn't public API (and the field is private), so you'll have to copy it with a note that it needs to be kept in sync with the original (though I can't imagine that the format will change).

The use of `firstMatch` below means that it can match even if there's non-whitespace before the comment:
```
void main() { //@dart=2.2
}
```
but that isn't considered a version comment because the version comment has to come before any non-comment in the file. The use of `firstMatch` also means that it could match text in a string literal. Both of those are unlikely possibilities, but there is a slightly different approach that would solve those cases and be more performant when dealing with larger files.

The AST structure provides access to the underlying tokens, including comments. In particular you can ask any node for the comment tokens that proceed it. The class `LanguageVersionOverrideVerifier` uses this to limit the search to only those comments that appear before the first non-comment token. Consider using a similar approach.

File: test_data/integration/null_safety/src/null_safety.dart
Comment (bwilkerson) [inline test_data/integration/null_safety/src/null_safety.dart:5]: I'd recommend testing boundary conditions, so consider using `2.11` instead. We should also have a test with `2.12` to ensure that we're not generating a lint in that case.

**Review by bwilkerson — COMMENTED** [review]
Thanks!

You should be able to handle the pubspec case by overriding `PubspecVisitor getPubspecVisitor()` in your subclass of `LintRule`. I haven't verified that it will give you access to everything you need, but we can extend it if necessary.

File: lib/src/rules/null_safety.dart
Comment (asashour) [reply lib/src/rules/null_safety.dart:34]: I see, it is now `enable_null_safety`. Another suggestion is `ensure_sound_null_safety`.

File: lib/src/rules/null_safety.dart
Comment (asashour) [reply lib/src/rules/null_safety.dart]: I see, in the previous code there was `.replaceAll(' ', '')` to remove all whitespaces before using the `RegExp`.

Anyhow, it is now changed to the regular expression from `LanguageVersionOverrideVerifier` which contains the initial line `^`, I believe that would make things better, since it will always consider line beginning.

Also, the comments of the first token are iterated.

Regarding the `PubspecVisitor`, by naked eye, I couldn't see something related to the `environment` or `sdk`, and since this is in `analyzer`, how about restricting this PR scope to the dart files. This `PubspecVisitor` seems to have dependency elsewhere.

Comment (bwilkerson) [reply lib/src/rules/null_safety.dart]: I missed the line to remove spaces, thanks.

Yes, I'm totally fine with merging this as-is and adding the pubspec handling later. I don't see support for the sdk version either, so we'll need to add that.

**Review by bwilkerson — APPROVED** [review]

**Comment by bwilkerson** [comment]
@pq Do we want to merge this now (well, after the conflict is resolved) and risk it being a breaking change to have additional checks, or would it be better to merge this after the next roll into the SDK so that it can't be used yet? A third option would be to mark it as experimental until the pubspec checking has been added. (FYI, Flutter is interested in this lint.)

File: lib/src/rules/enable_null_safety.dart
Comment (pq) [inline lib/src/rules/enable_null_safety.dart:19]: Perhaps:

```
a() {

}
```

?

File: lib/src/rules/enable_null_safety.dart
Comment (pq) [inline lib/src/rules/enable_null_safety.dart:25]: And,

```
b() {

}
```

?

File: lib/src/rules/enable_null_safety.dart
Comment (pq) [inline lib/src/rules/enable_null_safety.dart:40]: I don't know how expensive the comment parsing below is but I wonder if it's worth bailing if non-nullability is enabled?

```dart
    if (context.isEnabled(Feature.non_nullable)) {
      return;
    }
```

Comment (bwilkerson) [reply lib/src/rules/enable_null_safety.dart:40]: The comments have already been parsed and are in the AST so it's fairly inexpensive to iterate over them, but it wouldn't hurt to check.

Comment (pq) [reply lib/src/rules/enable_null_safety.dart:40]: 👍 

More specifically I mean the cost of the **_regexps_**.

Comment (asashour) [reply lib/src/rules/enable_null_safety.dart:40]: Thanks for hinting.

```dart
if (context.isEnabled(Feature.non_nullable)) {
  return;
}
```

doesn't work, giving all cases are ok (2 should fail).

On the other hand, there was also a suggestion to do something like:

```dart
if (line.replaceAll(' ', '').startsWith('//@dart=')) {
    // parse the version
}
```

if there is a concern about `regExp`.

**Comment by srawlins** [comment]
I think it would be super good to land this. Would you mind fixing the merge conflict, @asashour ?

**Comment by srawlins** [comment]
@bwilkerson @pq this is the rule that bans `// @dart=2.9`, etc. Likely not needed for Dart >= 3.0, but maybe extra useful before then.

File: lib/src/rules/enable_null_safety.dart
Comment (bwilkerson) [inline lib/src/rules/enable_null_safety.dart:54]: I didn't find any documentation that explicitly mentions it, but I think it's valid to have both a script tag and a language version override comment in the same file, and I'm guessing that the comment follows the script tag.

That's an unlikely corner case, but if I'm right then the `beginToken` might be a script tag token, in which case you would need to skip to the next token before getting the `precedingComments`.

File: lib/src/rules/enable_null_safety.dart
Comment (bwilkerson) [reply lib/src/rules/enable_null_safety.dart:39]: It might not be too bad, given that we only need to look at the comments before the first non-comment token in the file (modulo the corner case below). The performance is then dependent on the length of the file header and library documentation comments.

File: test/integration/enable_null_safety.dart
Comment (bwilkerson) [inline test/integration/enable_null_safety.dart:15]: I'd rather have these written using `reflective_test_loader`, but doing so shouldn't hold up merging the PR.

**Review by bwilkerson — APPROVED** [review]

**Review by bwilkerson — APPROVED** [review]

**Review by pq — APPROVED** [review]
👍

**Comment by pq** [comment]
I'd like to hold off on landing this until after the currently in progress roll is finished. (See: https://dart-review.googlesource.com/c/sdk/+/264961)

Thanks for your patience!

**Pull request merged**
